### PR TITLE
fix(auth): confine workspace-scoped API key to its workspace (#963)

### LINKED ISSUE #963: security/needs-triage: workspace-scoped API key not confined to its workspace on context read
## Severity: needs-triage (potential Critical — security) — spec confirmation first

A workspace-scoped MCP API key does not appear to be confined to its workspace on context reads. `_resolve_context_for_read` (`backend/src/mcp_server/tools/_helpers.py:181`) calls `PermissionService.resolve_context_for_workspace_read(user_id, context_id, required_role)` — it passes the **user_id and the target context**, but not the `session.workspace_id` carried by the API key. `resolve_context_for_workspace_read` (`backend/src/services/permission_service.py:296+`) authorises purely on the caller's **membership in the context's owning workspace**.

Net effect: a key minted for workspace A can read/write a context owned by workspace B whenever the key's user is also a member of B. The key's workspace scope is not enforced at the context layer (transfer path confirmed in source).

### Question / Fix
Is a workspace-scoped key intended to confine access to that workspace? If yes: when `session.workspace_id` is set, assert `context.workspace_id == session.workspace_id` at the MCP dispatch / `_resolve_context_for_read` layer and raise `_ContextNotFoundError` on mismatch. If the cross-workspace behaviour is by design (key = user-scoped), document it explicitly.

---
_From the 2026-06-10 cross-repo ecosystem audit. Transfer path verified against source; severity noted per item._

### PR DESCRIPTION
**Pull request by JFK**
Closes #963

## Summary
- A workspace-scoped API key authorized context reads purely on the caller's membership in the context's **owning** workspace, never comparing the key's own workspace. A key minted for workspace A could read/write a context owned by B whenever the key's user was also a member of B — the key scope silently downgraded to user scope (CWE-639 class).
- Enforce confinement in the single service-layer chokepoint `PermissionService.resolve_context_for_workspace_read` via a new `key_workspace_id` param: when supplied, require `context.workspace_id == key_workspace_id`, else raise the uniform `NotFoundException` (404). Skipped when `None` (OAuth2/session auth carries no key scope, per #889) so session/UI reads are unaffected.
- Wired from both API-key-reachable surfaces: MCP (`_resolve_context_for_read` forwards `workspace_id` across 12 read/write call sites) and REST (`agent_state` ×4 + `memory` list pass `api_key_workspace_id`).

## Implementation notes
- **Scope note**: this PR is broader than the issue title (which named the MCP read path). Verification during review found the *same* bypass on the REST read path, so the fix is placed at the shared service chokepoint and closes both surfaces of the vulnerability class rather than half-closing it.
- `graph.py` is intentionally **not** wired: it uses `SessionUser` (session-cookie only), so an API key cannot reach it — there is no key scope to confine.
- Both sides of the comparison are real `uuid.UUID` (`Context.workspace_id` mapped as UUID; `api_key_workspace_id` from `APIKey.workspace_id`); no str/UUID coercion hazard.
- Tests: service enforcement (3) + MCP forwarding/e2e (3, incl. `handle_feedback` wiring guard) + REST forwarding (agent_state); existing call-signature pins updated. 1167 (services + mcp_server) + 44 (route) pass; ruff clean.
- Follow-up (non-blocking, from gate2 qa-lead): add direct forwarding tests for the write path (`handle_remember`) and the REST `memory` list route — both are same-pattern duplicates of already-tested sites.

## Pre-PR review summary
- gate2 mode: advisor-only
- audit: skipped
- binary_gate: (none)
- cso: green
- qa-lead: green
- cto: green
- gate1: green via /ask (routed to CSO)
- review provider: code-review

Full reviews are saved in the plugin cache:
- ~/.claude/cache/gh-issue-driven/963-fix-workspace-key-context-isolation.gate1.md
- ~/.claude/cache/gh-issue-driven/963-fix-workspace-key-context-isolation.gate2.md

🤖 Generated via /gh-issue-driven:ship

### CODE REVIEW

File: backend/src/mcp_server/tools/_helpers.py [hunk 1]
```diff
@@ -197,13 +198,30 @@ async def _resolve_context_for_read(
...
     try:
+        # Issue #963: forward the key's workspace scope so confinement is
+        # enforced in the single service-layer chokepoint
+        # (resolve_context_for_workspace_read), shared with the REST read paths.
         return await PermissionService(db).resolve_context_for_workspace_read(
-            user_id=user_id, context_id=context_id, required_role=required_role
+            user_id=user_id,
+            context_id=context_id,
+            required_role=required_role,
+            key_workspace_id=workspace_id,
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/mcp_server/tools/_helpers.py]: Confirmed — this is a real bug, thank you. transport.py:419-423 sets `workspace_id = api_key_workspace_id if api_key_workspace_id else _get_user_workspace_id(user_id)`, and `_get_user_workspace_id` returns `User.current_workspace_id`. So for OAuth2/session and **global** (non-workspace-scoped) API keys the handler `workspace_id` is the user's *current* workspace (non-None), and forwarding it as `key_workspace_id` over-confines those callers — contradicting the intended 'session/UI reads unaffected'. The REST side is correct (it uses `api_key_workspace_id` directly). Fix: plumb `api_key_workspace_id` through MCP dispatch (execute_tool_call → handlers) and forward THAT as `key_workspace_id`, keeping `workspace_id` for rate-limit/telemetry — exactly your recommendation. Leaving this thread open until the fix lands.

File: backend/src/mcp_server/tools/memory.py [hunk 2]
```diff
@@ -281,7 +281,9 @@ async def handle_recall_upcoming(
             current_context_id = _resolve_context_id(args["context_id"])
             # Read path: uniform context_not_found on any deny (CWE-639 / OWASP
             # A01), mirroring handle_recall.
-            current_context = await _resolve_context_for_read(db, user_id, current_context_id)
+            current_context = await _resolve_context_for_read(
+                db, user_id, current_context_id, workspace_id=workspace_id
+            )
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/mcp_server/tools/memory.py]: Confirmed — the MCP write handlers (handle_remember:72, handle_update_memory:174, handle_forget:644) resolve via `_resolve_context` (ContextService.get_context, membership-only), so they are NOT confined — the #963 class remains open on the MCP write path. This will be fixed together with the key-scope plumbing from the thread on _helpers.py: the write paths will thread the API key's workspace scope into the service-layer chokepoint (migrate to `_resolve_context_for_read(..., key scope)` or extend `_resolve_context` to enforce it). Leaving open until fixed.

File: backend/src/api/routes/memory.py [hunk 3]
```diff
@@ -676,7 +676,9 @@ async def list_memories(
             # Raises 404 on non-existent context, non-member, or private-context
             # non-creator (CWE-639 uniform disclosure). Matches the graph routes.
             context = await PermissionService(db).resolve_context_for_workspace_read(
-                user_id=user_id, context_id=context_id
+                user_id=user_id,
+                context_id=context_id,
+                key_workspace_id=user.get("api_key_workspace_id"),
             )
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/api/routes/memory.py:682]: Agreed — adding a /memory/list test that passes a user dict with a non-None `api_key_workspace_id` and asserts `resolve_context_for_workspace_read(..., key_workspace_id=<ws>)`, mirroring the agent_state wiring guards. Will land with the key-scope plumbing fix.

File: backend/src/mcp_server/tools/_helpers.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/mcp_server/tools/_helpers.py]: ✅ Fixed in db5f4e40. Switched MCP to a per-request contextvar carrying the PURE api_key_workspace_id (None unless a workspace-scoped key). transport.py sets it after auth; _resolve_context_for_read reads it instead of the conflated handler workspace_id. OAuth2/session/global-key callers now have key scope None → not confined. The 12 handler call-site params were removed.

File: backend/src/mcp_server/tools/memory.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/mcp_server/tools/memory.py]: ✅ Fixed in db5f4e40. _resolve_context (the write-path resolver used by handle_remember/update_memory/forget) now enforces the same confinement after ContextService.get_context, using the per-request key-scope contextvar → uniform _ContextNotFoundError on mismatch. MCP writes are now covered alongside reads.

File: backend/src/api/routes/memory.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/api/routes/memory.py:682]: ✅ Done in db5f4e40. Added a /memory/list test passing a user dict with a non-None api_key_workspace_id and asserting resolve_context_for_workspace_read(..., key_workspace_id=<ws>).

File: backend/src/mcp_server/tools/_helpers.py [hunk 4]
```diff
@@ -169,14 +194,36 @@ async def _resolve_context(
...
+        logger.warning(
+            "context_write_denied",
+            extra={
+                "reason": "key_workspace_mismatch",
+                "context_id": str(context_id),
+                "context_workspace_id": str(context.workspace_id),
+                "key_workspace_id": str(key_workspace_id),
+                "user_id": user_id,
+            },
+        )
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/mcp_server/tools/_helpers.py:223]: ✅ Fixed in 16c94831. Switched to %-args in the log message (matching this module's _log_tool_usage style) so reason/context_id/context_workspace_id/key_workspace_id/user_id actually render under the stdlib "%(message)s" formatter — extra={...} was silently dropped.

File: backend/src/services/permission_service.py [hunk 5]
```diff
@@ -268,6 +269,15 @@ async def resolve_context_for_workspace_read(
...
                 ``owner``.
+            key_workspace_id: Issue #963. The caller's *key* workspace scope —
+                ``session.workspace_id`` for MCP (carried by a workspace-scoped
+                API key) or ``api_key_workspace_id`` for REST. When supplied,
+                confine the read to it: a key minted for workspace A must not
+                reach a context owned by B even when the user is also a member
+                of B (membership above is necessary but not sufficient). Pass
+                ``None`` for session/OAuth callers, which carry no key scope
+                (#889) and are governed by membership alone. A mismatch raises
+                the same uniform 404 as every other deny.
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/services/permission_service.py]: ✅ Fixed in c085c442. Rewrote the key_workspace_id docstring to state it is the PURE api-key workspace scope (None unless a workspace-scoped key) and explicitly NOT session.workspace_id / the conflated current workspace — MCP passes api_key_workspace_id via the per-request contextvar, REST passes user['api_key_workspace_id'].

File: backend/src/services/permission_service.py [hunk 6]
```diff
@@ -365,6 +375,22 @@ async def resolve_context_for_workspace_read(
             )
             raise NotFoundException("Context", str(context_id))
 
+        # Issue #963: confine a workspace-scoped caller (API key carrying
+        # key_workspace_id) to its own workspace. Membership above is necessary
+        # but not sufficient — a key for workspace A must not reach a context
+        # owned by B even when the user is also a member of B. Skip when
+        # key_workspace_id is None (session/OAuth carries no key scope, #889).
+        if key_workspace_id is not None and context.workspace_id != key_workspace_id:
```
[the remark this answers is not in the document]
Comment (JFK) [reply backend/src/services/permission_service.py]: ✅ Fixed in c085c442. Moved the mismatch check to immediately after the Context load, before check_workspace_access — a cross-workspace key now short-circuits to the uniform 404 without the membership-lookup query (no DB-load amplification under probing). Behavior is unchanged (all denials remain uniform 404).
